### PR TITLE
[SEMI-MODULAR] (mostly) fixes AI improvement PR code

### PR DESCRIPTION
**Pull request by Seris02**
## About The Pull Request

#7063 made it so all AIs could hear all emotes from all ai_eye objects, as well as some other assorted bugs

the code was utter shit too, whoever wrote that code, please, make sure you look for variables to see if it can be done in a better way other than massive for loops

## Why It's Good For The Game

bug fix, code improvement

## Changelog
:cl:
fix: AIs will no longer hear all emotes from all ai_eye cameras
/:cl:

### CODE REVIEW

**Comment by FlamingLily** [comment]
Multikey when testing is easy, first join the server, then close the pager/byond hub, then join the server again, you'll join as a guest

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 1]
```diff
@@ -40,22 +40,13 @@
 
 	var/list/heard = get_hearers_in_view(7, get_top_level_mob(src.mob))
 
-	//SKYRAT EDIT ADDITION BEGIN - AI QoL
-	var/ai_heard_it = FALSE
 	//so the ai can post looc text
 	if(istype(mob,/mob/living/silicon/ai))
```
Comment (FlamingLily) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm:44]:
```suggestion
	if(istype(mob,/mob/living/silicon/ai))
	    var/mob/living/silicon/ai/ai = mob
```

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 2]
```diff
@@ -40,22 +40,13 @@
...
-	for(var/mob/camera/ai_eye/eye_of_ai as anything in GLOB.aiEyes)
-		for(var/mob/creature in heard) //checks if an ai has already heard it
-			if(istype(creature, /mob/living/silicon/ai))
-				ai_heard_it = TRUE
-		if(ai_heard_it == FALSE) //i dont wanna run an entire fuckin for loop if it turns out the ai can already see the text
-			for(var/mob/creature in heard)
-				if(creature == eye_of_ai)
-					heard += get_top_level_mob(src.mob)
-	//SKYRAT EDIT ADDITION END - AI QoL
+	for(var/mob/ai in GLOB.ai_list)
```
Comment (FlamingLily) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]:
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
```

File: code/datums/emotes.dm [hunk 3]
```diff
@@ -130,13 +130,8 @@
 
 	//SKYRAT EDIT ADDITION BEGIN - AI QoL
 	for(var/mob/ai in GLOB.ai_list)
```
Comment (FlamingLily) [inline code/datums/emotes.dm]:
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
```

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm (hunk 1 above)
Comment (FlamingLily) [reply modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm:44]: Beware indentation, I had to use spaces

**Review by FlamingLily — CHANGES_REQUESTED** [review]
Shiz broke, this should fix, but no promises, is webedit after all

File: code/datums/emotes.dm (hunk 3 above)
Comment (FlamingLily) [reply code/datums/emotes.dm]: I'ma be real I actually don't know the difference between using as anything or not but the last time I had to do smth like this I was told to use as anything so there

**Comment by SpaceVampire** [comment]
Wasn't it #7063 that did that? I think you might have mentioned the wrong PR up above.

File: code/datums/emotes.dm [hunk 4]
```diff
@@ -129,14 +129,9 @@
...
 	//SKYRAT EDIT ADDITION BEGIN - AI QoL
-	for(var/mob/ai in GLOB.ai_list)
-		//ai.show_message("<span class='emote'>[!(ai.stat == DEAD)]</span>")
-		var/aiEye_turf
-		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
-			aiEye_turf = get_turf(AI_eye)
-		if(!ai.client)
-			continue
-		if(!(ai.stat == DEAD) && ai.client && (get_dist(user_turf, aiEye_turf)<8))
+	for(var/mob/living/silicon/ai/ai in GLOB.ai_list)
```
Comment (tf-4) [inline code/datums/emotes.dm]:
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
```
Shouldn't be able to be any other type here AFAIK.

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 5]
```diff
@@ -40,22 +40,13 @@
...
-	for(var/mob/camera/ai_eye/eye_of_ai as anything in GLOB.aiEyes)
-		for(var/mob/creature in heard) //checks if an ai has already heard it
-			if(istype(creature, /mob/living/silicon/ai))
-				ai_heard_it = TRUE
-		if(ai_heard_it == FALSE) //i dont wanna run an entire fuckin for loop if it turns out the ai can already see the text
-			for(var/mob/creature in heard)
-				if(creature == eye_of_ai)
-					heard += get_top_level_mob(src.mob)
-	//SKYRAT EDIT ADDITION END - AI QoL
+	for(var/mob/living/silicon/ai/ai in GLOB.ai_list)
```
Comment (tf-4) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]:
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
```

**Comment by Gandalf2k15** [comment]
anythinng

File: code/datums/emotes.dm [hunk 6]
```diff
@@ -129,14 +129,9 @@
...
 	//SKYRAT EDIT ADDITION BEGIN - AI QoL
-	for(var/mob/ai in GLOB.ai_list)
-		//ai.show_message("<span class='emote'>[!(ai.stat == DEAD)]</span>")
-		var/aiEye_turf
-		for(var/mob/camera/ai_eye/AI_eye as anything in GLOB.aiEyes)
-			aiEye_turf = get_turf(AI_eye)
-		if(!ai.client)
-			continue
-		if(!(ai.stat == DEAD) && ai.client && (get_dist(user_turf, aiEye_turf)<8))
+	for(var/mob/living/silicon/ai/ai as anythinng in GLOB.ai_list)
```
Comment (FlamingLily) [inline code/datums/emotes.dm]:
```suggestion
	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
```
anythinng

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm (hunk 1 above)
Comment (FlamingLily) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm:44]:
```suggestion
	if(istype(mob,/mob/living/silicon/ai))
	    var/mob/living/silicon/ai/ai = mob
```
you marked this as resolved but didn't actually do anything and is still throwing errors

File: modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm [hunk 7]
```diff
@@ -40,22 +40,13 @@
...
-		for(var/mob/creature in heard) //checks if an ai has already heard it
-			if(istype(creature, /mob/living/silicon/ai))
-				ai_heard_it = TRUE
-		if(ai_heard_it == FALSE) //i dont wanna run an entire fuckin for loop if it turns out the ai can already see the text
-			for(var/mob/creature in heard)
-				if(creature == eye_of_ai)
-					heard += get_top_level_mob(src.mob)
-	//SKYRAT EDIT ADDITION END - AI QoL
+	for(var/mob/living/silicon/ai/ai as anything in GLOB.ai_list)
+		if(ai.client && !(ai in heard) && ai.eyeobj in heard)
```
Comment (FlamingLily) [inline modular_skyrat/modules/verbs/code/modules/client/verbs/looc.dm]:
```suggestion
		if(ai.client && !(ai in heard) && (ai.eyeobj in heard))
```

**Review by FlamingLily — CHANGES_REQUESTED** [review]

**Comment by Seris02** [comment]
my brain has just
completely died for this PR
fuck

**Comment by tf-4** [comment]
link issue https://github.com/Skyrat-SS13/Skyrat-tg/issues/7250

**Pull request merged**